### ISSUE TITLE
Special characters display problem

### ISSUE THREAD

**Issue by nguyend** [nlplab/brat#1115]
Hi, 
I tried to annotate some tweets but it seems the client did not displayed the "emotion" character well.
All the spaces are missing after this 😂 (attached snapshot).
And the page is not responding with these 01F44A, 01F4AA,01F48B

Please suggest how to deal with these character either remove or keep it.
Thanks and Regards,

![image](https://cloud.githubusercontent.com/assets/6589663/4571717/3be038cc-4f7c-11e4-96b1-ee4059cdea30.png)

**Comment by kennyjoseph** [comment]
Hey @nguyend, I'm seeing the same issues - it appears to happen with emoticons, particularly with Twitter data. Here's an example text that causes the issue:

```
@kenziedannielle 😂😂 yes the Shane's in the boro . Come see me tonight or tomorrow night for dinner .
```

Did you ever resolve this? FTR @ninjin I can replicate this on Safari, Chrome and Firefox (latest versions)

**Comment by ninjin** [comment]
My guess is that the width of the glyphs are misreported by the browser and screws up the layout algorithm.  Any better guess @amadanmath?

**Comment by amadanmath** [comment]
It seems that the display bug has already been fixed in master branch, as I can't reproduce it.

However, there is a more insidious bug here: all annotations for text that follows any astral characters will be wrong (offset by one for each preceding astral character). The fact that it even works is an amazing coincidence owing to the fact that Python2 and JavaScript both don't know how to handle astral characters, and break in the exactly same way: by treating them as if they were two characters. Thus, if you annotate `"yes"` in `"😂😂 yes"`, the standoff should be 3-6 (the first three characters being emoji, emoji and space); but it is recorded as 5-8 (the first five characters being emoji-part-1, emoji-part-2, emoji-part-1, emoji-part-2 and space).

There are three ways around it that I can see, and none are pretty:

* Fix annotations by postprocessing (deduce 1 from each standoff index for each astral character before it)
* Replace all astral characters by basic plane stand-ins
* Replace any use of `length` property, `substr`, `substring`, `charAt` and `charCodeAt` methods (and Python equivalents) on strings with unicode-compliant versions (huge task, I'm very unlikely to do it).

Now that I think about it, this should probably be a separate issue...

**Comment by kennyjoseph** [comment]
Yikes. @amadanmath, thank you very much for the explanation, and yes, you're right about me talking about the latest tagged version.  For right now, I'm just replacing them, as it wasn't really necessary for my task. However, I can open the separate issue if you like.

**Comment by amadanmath** [comment]
More info: Python2 apparently can be built using ucs2 ("narrow build") or ucs4 ("wide build"). The latter uses four bytes for each character, but supports astral characters (and probably bugs up brat; I haven't tried). It seems wide is the default on Linuxen, and narrow on OSX and Windows. Yay.

**Comment by spyysalo** [comment]
@amadanmath : if switching for a wide python build were to fix the server side, could JS `string.length` (etc.) be monkey-patched to switch in a function that supports this on the client?

**Comment by parthjoshi2007** [comment]
Hi, any progress on this issue? I'm facing this problem on a Linux machine with Python2. Also could you please add the fact that emoticons break the text rendering somewhere in the documentation because for some use cases like mine with Twitter data, being able to annotate emoticons and other unicode characters is a requirement? I loved Brat during testing and hence selected it as my annotation tool of choice and having integrated into my workflow, now halfway through my annotation exercise, I'm having to create the .ann files by hand. 😢

**Comment by ThiagoSousa** [comment]
I also have the same problem. Is there any solution to this? In fact, it seems that it happens in a linux server (I'm using CentOS 6.9). On my local OSX High Sierra I have no problem rendering the emojis.

**Comment by kovvalsky** [comment]
Using the latest master `7a58bbe` and texts following emojis are overlapping in chromium 81.0. :( 
Though [this comment](https://github.com/nlplab/brat/issues/1115#issuecomment-128594508) suggests that the display bug is fixed (not offset bug).

**Comment by amadanmath** [comment]
Right... As I said in that comment, Python 2 and JavaScript both fail in the same way, so it's a lucky coincidence that it worked. Python 2 version has since moved to the `python2` branch (given that Python 2 has reached end of life), and `master` was converted to Python 3. So the server started counting correctly, but the client is still miscounting.

I will try to see if I can get JavaScript to do it correctly next week (per the third bullet above), but I'm not promising anything. I'm also worried about speed impact, but that may be premature.

(BTW JavaScript `String.prototype.length` is kind of magic, and cannot be monkey-patched. So all uses of `.length` would need to be checked and changed.)

**Comment by amadanmath** [comment]
I managed to hack something together. There should be no impact on documents without astral characters (such as "😂"); if you're interested in astral character support and have data to test with, I would be grateful if you got the new `master` and tell me if anything breaks.

**Comment by kovvalsky** [comment]
Thank you for your response @amadanmath 
I updated my copy according to the `master` but getting the error in logs:
```assert isfile(SIMSTRING_EXECUTABLE), "Error: simstring not found (Hint: set SIMSTRING_EXECUTABLE in config.py
```
Is it mandatory to have `SIMSTRING_EXECUTABLE` set in `config.py`?

**Comment by kovvalsky** [comment]
After installing `simstring`, I tried emojis with the current `master`.
- The rendering of the emojis is fine, the subsequent characters are not overlapping anymore.
- The selection is tricky, and I guess your commit didn't intend to solve this.
See the gif:
![Alt Text](https://naturallogic.pro/pub_img/emoji_selection_in_brat.gif)
The sample string: `0😔2😔4😷6😂8👉0`

**Comment by amadanmath** [comment]
Thank you so much for the feedback. I was working on normalization before, and ended up requiring it whether or not it is actually used. That should be fixed now.

Also, fixed the detection of offsets when a new span is selected. If you don't mind, please pull and check again if you find your issues resolved (and also if you find some other ones :P )

**Comment by tonifuc3m** [comment]
Hi,
I am using the last committed version of the master branch and, as @kovvalsky reported, the emoji display works fine, but the selection does not. The same happens if we have other weird unicode characters such as **bolded** words.

![simplescreenrecorder-2020-11-09_10 49 522](https://user-images.githubusercontent.com/46200970/98526581-36762900-227a-11eb-831d-39a0cd34e5b6.gif)

Has there been some work to fix this?

I attach the example txt file. 
[testing.txt](https://github.com/nlplab/brat/files/5509569/testing.txt)

**Comment by amadanmath** [comment]
Thanks for the report, you're the first to say something after I hacked
this.
I think I fixed it; can you pull and check, please?

Goran

**Comment by tonifuc3m** [comment]
It seems to be working perfectly after the git pull. Thanks for the quick reply!
